cmd/tailscale/cli/ffcomplete/internal/complete.go

```diff
@@ -0,0 +1,254 @@
+		emitFlag = false
+		emitArgs = false
+
+		dashFlag, completeVal, _ := strings.Cut(completeArg, "=")
+		_, f := cutDash(dashFlag)
+		flag := cmd.FlagSet.Lookup(f)
+		if flag != nil {
+			if comp := CompleteFlags[flag]; comp != nil {
+				// Complete custom flag values.
+				var err error
+				words, dir, err = comp([]string{completeVal})
+				if err != nil {
+					return nil, 0, fmt.Errorf("completing %s flag %s: %w", cmd.Name, flag.Name, err)
+				}
+			} else if isBoolFlag(flag) {
+				// Complete true/false.
+				for _, vals := range [][]string{
+					{"true", "TRUE", "True", "1"},
+					{"false", "FALSE", "False", "0"},
+				} {
+					for _, val := range vals {
+						if strings.HasPrefix(val, completeVal) {
+							words = append(words, val)
+							break
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Complete '-flag...'.
+	if emitFlag {
+		used := make(map[string]struct{})
+		cmd.FlagSet.Visit(func(f *flag.Flag) {
+			used[f.Name] = struct{}{}
+		})
+
+		cd, cf := cutDash(completeArg)
+		cmd.FlagSet.VisitAll(func(f *flag.Flag) {
+			if !strings.HasPrefix(f.Name, cf) {
+				return
+			}
+			// Skip flags already set by the user.
+			if _, seen := used[f.Name]; seen {
+				return
+			}
+			// Suggest single-dash '-v' for single-char flags and
+			// double-dash '--verbose' for longer.
+			d := cd
+			if (d == "" || d == "-") && cf == "" && len(f.Name) > 1 {
+				d = "--"
+			}
+			if descs {
+				_, usage := flag.UnquoteUsage(f)
+				usage = strings.TrimPrefix(usage, "HIDDEN: ")
+				if usage != "" {
+					words = append(words, d+f.Name+"\t"+usage)
+					return
+				}
+			}
+			words = append(words, d+f.Name)
+		})
+	}
+
+	if emitArgs {
+		// Complete 'sub...'.
+		for _, sub := range cmd.Subcommands {
+			if strings.HasPrefix(sub.Name, completeArg) {
+				if descs {
+					if help := strings.TrimPrefix(sub.ShortHelp, "HIDDEN: "); help != "" {
+						words = append(words, sub.Name+"\t"+help)
+						continue
+					}
+				}
+				words = append(words, sub.Name)
+			}
+		}
+
+		// Complete custom args.
+		if comp := CompleteCmds[cmd]; comp != nil {
+			w, d, err := comp(append(args, completeArg))
+			if err != nil {
+				return nil, 0, fmt.Errorf("completing %s args: %w", cmd.Name, err)
+			}
+			dir = d
+			words = append(words, w...)
+		}
+	}
+
+	// Strip any descriptions if they were suppressed.
+	if !descs {
+		for i := range words {
+			words[i], _, _ = strings.Cut(words[i], "\t")
+		}
+	}
+	return words, dir, nil
+}
+
+// splitFlagArgs separates a list of command-line arguments into arguments
+// comprising flags and their values, preceding arguments to be passed to the
+// command. This follows the stdlib 'flag' parsing conventions. If the final
+// argument is a flag name which takes a value but has no value specified, it is
+// omitted from flagArgs and argArgs and instead returned in needValue.
+func splitFlagArgs(fs *flag.FlagSet, args []string) (flagArgs, argArgs []string, flagNeedingValue string) {
```

Comment:
Could fuzz this function to make sure the splitting is accurate? The fuzz case would be a `[]string args`. On one side, we do flag.Parse() to get the expected results; then compare with the other side where we `flag.Parse(splitFlagArgs(args))` the Visitd()'d flags are the same, and remaining args matches expected Args().

Reply:
```
$ go test -fuzz Split ./cmd/tailscale/cli/ffcomplete/internal/
fuzz: elapsed: 0s, gathering baseline coverage: 0/19 completed
failure while testing seed corpus entry: FuzzSplitFlagArgs/c64b69bf2c432100
fuzz: elapsed: 0s, gathering baseline coverage: 0/19 completed
--- FAIL: FuzzSplitFlagArgs (0.02s)
    --- FAIL: FuzzSplitFlagArgs (0.00s)
        complete_test.go:262: "-0" => (-flag.Parse +SplitFlagArgs):
              internal_test.res{
            - 	Err:     e"flag provided but not defined: -0",
            + 	Err:     nil,
              	Visited: nil,
            - 	Args:    []string{},
            + 	Args:    []string{"-0"},
              }
```